Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -742,12 +743,112 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
         try? self.init(throwsIfInvalid: description)
     }
 
-    // TODO: SWIFT-1405: add options to description
     public var description: String {
-        var des = ""
-        des += "\(self.scheme)://"
-        des += self.hosts.map { $0.description }.joined(separator: ",")
-        return des
+        var uri = ""
+        uri += "\(self.scheme)://"
+        if let username = self.credential?.username {
+            uri += username.getPercentEncoded(withAllowedCharacters: .urlUserAllowed)
+            if let password = self.credential?.password {
+                uri += ":" + password.getPercentEncoded(withAllowedCharacters: .urlPasswordAllowed)
+            }
+            uri += "@"
+        }
+        uri += self.hosts.map { $0.description }.joined(separator: ",")
+        // A trailing slash in the connection string is valid so we can append this unconditionally.
+        uri += "/"
+        if let defaultAuthDB = self.defaultAuthDB {
+            uri += defaultAuthDB.getPercentEncoded(withAllowedCharacters: .urlQueryAllowed)
+        }
+        uri += "?"
+        uri.appendOption(name: .appName, option: self.appName)
+        uri.appendOption(name: .authMechanism, option: self.credential?.mechanism?.description)
+        uri.appendOption(name: .authMechanismProperties, option: self.credential?.mechanismProperties?.map {
+            var property = $0.key + ":"
+            switch $0.value {
+            case let .string(s):
+                property += s
+            case let .bool(b):
+                property += String(b)
+            // the possible values for authMechanismProperties are only strings and booleans
+            default:
+                property += ""
+            }
+            return property
+        }.joined(separator: ","))
+        uri.appendOption(name: .authSource, option: self.credential?.source.flatMap {
+            // if authSource is equal to either of these values, it was set as a default rather than by an authSource
+            // URI option
```

Comment:
The default source depends on the auth mechanism, so we can't always just check for `"admin"` here (e.g. for x509 its `"$external"`). If we want to strip out the defaults, we'll need to confer with the mechanism first. Alternatively, we could just always put the source in.

Reply:
updated to check against the mechanism's default

Reply:
sorry for the back and forth, but thinking about it more I wonder if Patrick's suggestion to always put the source in would be better. it seems like it might be less surprising to populate a connection string with a default value (that a user left unspecified) than it would be to strip out a user-provided value just because it was the default. thoughts?

Reply:
hm, I guess I was assuming that users wouldn't be likely to specify a default explicitly but I suppose that's not too unreasonable? in that case it's probably better to err on the side of always adding an authsource, I'll update

Reply:
The non-SRV URIs provided by the Atlas UI actually specify the authSource explicitly, so this could be somewhat common. The SRV URIs also specify it "explicitly" via TXT record, though that is never really visible to the user.